### PR TITLE
[FSDP][dtensor] use _StridedShard to represent nested sharding for correct full_tensor() result

### PR DESCRIPTION
**Pull request by XilunWu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130760
* #131408
* #132391
* #130239
* #126697

Fixes issue #129229 #129206 
**Summary**

1. Have `FSDP` choose `_StridedShard` placement for FSDP+TP sharding
2. Added a parity test to FSDP to ensure that FSDP+TP sharding (i.e. strided) and simply TP sharding (i.e. non-strided) has the same `full_tensor()` result
3. Re-enabled the tests that were disabled in #129519 

**test**
`pytest test/distributed/_composable/fsdp/`
`pytest test/distributed/_composable/test_composability/test_2d_composability.py`
`pytest test/distributed/checkpoint/fsdp/test_fsdp_dsd.py`

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @zhaojuanmao @mrshenli @rohan-varma @chauhang @LucasLLC @MeetVadakkanchery @mhorowitz @pradeepfn

Differential Revision: [D60606114](https://our.internmc.facebook.com/intern/diff/D60606114)

### CODE REVIEW

File: test/distributed/_composable/fsdp/test_fully_shard_state_dict.py [hunk 1]
```diff
@@ -104,10 +104,55 @@ def test_1d_state_dict_cpu_offload(self):
...
+            },
+        )
+        for mlp in model_2d:
+            fully_shard(mlp, mesh=dp_mesh)
+        fully_shard(model_2d, mesh=dp_mesh)
+
+        # state_dict parity check
+        for param_name, param_tp in model_tp.named_parameters():
+            param_2d = model_2d.get_parameter(param_name)
+            self.assertEqual(param_tp.full_tensor(), param_2d.full_tensor())
```
Comment (fegin) [inline test/distributed/_composable/fsdp/test_fully_shard_state_dict.py:152]: Should we also test `state_dict()` call? `state_dict()` will be the actual API users use.

**Comment by wanchaol** [comment]
https://hud.pytorch.org/pr/pytorch/pytorch/130760#27830275868

test failures looks related.

Also, I think we should also adjust `compute_offset_and_local_shape` to have StridedShard compute the correct offset? Once that's supported when we can remove the state_dict hook that throwing errors

**Comment by XilunWu** [comment]
@XilunWu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60606114).

File: torch/distributed/_tensor/placement_types.py [hunk 2]
```diff
@@ -799,6 +799,29 @@ def dim_map(self) -> List[int]:
                 r[shard_dim] = i
         return r
 
+    @property
+    def num_shards_map(self) -> List[int]:
```
Comment (wanchaol) [inline torch/distributed/_tensor/placement_types.py:803]: can this be live in the offset calculation PR?

**Review by wanchaol — APPROVED** [review]
Please add strided sharding uneveness check as inlined in the comments

**Review by fegin — COMMENTED** [review]
~~I'm confused, should https://github.com/pytorch/pytorch/pull/131408 be landed before or after this PR?~~

Discussed offline, the previous PR has assertRaise to catch the exception and this PR can remove the exception.

**Comment by XilunWu** [comment]
@XilunWu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60606114).

**Review by fegin — APPROVED** [review]

**Review by wz337 — APPROVED** [review]
LGTM!

File: torch/distributed/_composable/fsdp/_fsdp_param.py [hunk 3]
```diff
@@ -281,15 +286,33 @@ def _init_sharded_param(self, param: nn.Parameter, device: torch.device):
...
+                    else Shard(0)
+                ),
                 self._tp_spec.placements[0],
             )
             self._sharding_spec = DTensorSpec(
                 self._spmd_mesh,
                 self._spmd_placements,
                 tensor_meta=self._tp_spec.tensor_meta,
             )
+            # NOTE: FSDP+TP does not support uneven sharding for now
```
Comment (wanchaol) [inline torch/distributed/_composable/fsdp/_fsdp_param.py:303]: hmmm I think this check is too strict, we should ONLY check the tensor dimension 0 for uneven sharding when the TP sharding spec sharded on tensor dim 0, as:
1. DTensor supports uneven sharding computation for its normal case.
2. Only when strided sharding enabled we don't yet support uneven sharding, and the strided sharding only enabled when the TP's DTensor is sharded on tensor dim 0 (and FSDP2 shards on tensor dim 0 too, which is the case where StridedShard comes in)

So please fix this check to be more accurate

Comment (XilunWu) [reply torch/distributed/_composable/fsdp/_fsdp_param.py:303]: thanks for catching the mistake!! I added a check on `split_factor` and only perform the sharding evenness check when it's larger than `1` (i.e. tensor dim 0 has been sharded on TP mesh dim).
